### PR TITLE
Check that `#[may_dangle]` is properly applied

### PR DESCRIPTION
**Pull request by GoldsteinE**
It's only valid when applied to a type or lifetime parameter in `Drop` trait implementation.

Tracking issue: https://github.com/rust-lang/rust/issues/34761
cc https://github.com/rust-lang/rust/issues/34761#issuecomment-1208185551

### CODE REVIEW

File: compiler/rustc_passes/src/check_attr.rs [hunk 1]
```diff
@@ -1373,6 +1373,28 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
         }
     }
 
+    /// Checks if `#[may_dangle]` is applied to a lifetime or type generic parameter in `Drop` impl.
+    fn check_may_dangle(&self, hir_id: HirId, attr: &Attribute, target: Target) {
+        if matches!(target, Target::GenericParam(_))
```
Comment (compiler-errors) [inline compiler/rustc_passes/src/check_attr.rs:1378]: this first `matches` is redundant if we are just checking the node one line below (1379)

File: compiler/rustc_passes/src/check_attr.rs [hunk 2]
```diff
@@ -1373,6 +1373,28 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
...
+            && let hir::Node::GenericParam(param) = self.tcx.hir_node(hir_id)
+            && matches!(
+                param.kind,
+                hir::GenericParamKind::Lifetime { .. } | hir::GenericParamKind::Type { .. }
+            )
+            && matches!(param.source, hir::GenericParamSource::Generics)
+            && let parent_hir_id = self.tcx.parent_hir_id(hir_id)
+            && let hir::Node::Item(item) = self.tcx.hir_node(parent_hir_id)
+            && let hir::ItemKind::Impl(impl_) = item.kind
+            && let Some(trait_) = impl_.of_trait
```
Comment (compiler-errors) [inline compiler/rustc_passes/src/check_attr.rs:1388]:
```suggestion
            && let hir::Node::Item(hir::Item { kind: hir::ItemKind::Impl(impl_), .. }) = self.tcx.hir_node(parent_hir_id)
            && let Some(trait_) = impl_.of_trait
```

**Review by compiler-errors — APPROVED** [review]
Couple tiny tweaks, otherwise LGTM

File: compiler/rustc_passes/src/check_attr.rs (hunk 2 above)
Comment (GoldsteinE) [reply compiler/rustc_passes/src/check_attr.rs:1388]: This leads to somewhat ugly formatting:

```rust
&& let hir::Node::Item(hir::Item { kind: hir::ItemKind::Impl(impl_), .. }) =
    self.tcx.hir_node(parent_hir_id)
```

I think the current version is nicer, but I don’t feel that strong about it, so I can make this change if you prefer it.

File: compiler/rustc_passes/src/check_attr.rs (hunk 1 above)
Comment (GoldsteinE) [reply compiler/rustc_passes/src/check_attr.rs]: Fair, I removed this check.

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**